sam0: Implement watchdog driver

benpicco opened the pull request:
### Contribution description
Most SoCs include a Watchdog timer peripheral that will trigger a reset if it is not regularly updated.

This can be used to automatically reset a board that is stuck in an error condition.

Many Watchdogs will provide a window mode: Feeding the Watchdog has to occur before `period` is elapsed, but after `window` is over.
This is to prevent a live-lock type situation where the Watchdog is updated earlier than usual.

A Watchdog may also trigger an Early Warning Interrupt shortly before the Watchdog Period is elapsed.

This PR implements the Watchdog interface for the sam0 line of SoCs.

### Testing procedure

Run `tests/periph_wdt` ~~(depends on #11521)~~

Four tests should run during which the Watchdog will be fed five times, then a board reset should be triggered.

### Issues/PRs references
Tested on:
- [x] samd21
- [ ] saml11
- [x] saml21
- [x] samd5x

On samd5x the Watchdog is always clocked by OSCULP32K, no further configuration is needed / possible. The same might be true for SAML11 & SAML21.

There are still some issues, but I suppose it's good to get feedback early on.

Working:
- [x] Normal mode
- [x] Window mode
- [x] Early Warning mode

keestux commented:
On my SODAQ ONE (samd21g18a) board the test does not work. It prints nothing at all. Not even the "This is RIOT!" banner.

benpicco commented:
@keestux you need #11521 for the test to work.

keestux commented:
Why did you decide to use GCLK2 (which is potentially setup in rtc.c and rtt.c and cpu.c, and with different configurations)?
But I must say that I'm mostly confused with respect to GCLKs. Let me mention #7913, #7944 and #7969 which touch this subject.

benpicco commented:
I decided to use GCLK2 exactly because RTT and RTC also configure it to run at 1024 Hz, so if those modules are used no change in configuration is required.

I agree that the GCLK situation is quite a mess. We could require that certain GCLKs can always be expected to run at a certain frequency (looks like we need 48MHz for USB, 1kHz for RTC & RTT, maybe 1 MHz for timer).

Something in the likes of https://github.com/zephyrproject-rtos/zephyr/pull/15585 would certainly be nice to have.

keestux commented:
> @keestux you need #11521 for the test to work.

Ah, that works :-) Albeit that is a bit challenging to get it back in bootloader mode again (i.e. double tap must be done really quick).

aabadie commented:
Have you looked at #11252 before ? It's not the same CPU vendor but it also tries to provide a common API for watchdog and a test application. If I remember well, it was not that far from being merged.

Maybe there's room for joining the effort here. CC'ing @fjmolinas and @kaspar030

dylad commented:
@benpicco Now that #11521 is merged, could you rebase on top of master please ?
I'll test this PR on SAML21 and SAML11 as soon as this PR is ready :)

benpicco commented:
@dylad ok, rebased.

benpicco commented:
I'm fine with that!

An argument to keep `setup` and `enable` separate would be so that you could temporarily disable the watchdog and then enable it again without having to re-configure it.

 The separate `init` function is only there because I thought it's convention to have a parameter-less init function that can be called by auto_init. I can move that into `wdg_setup()`.

```
 void wdg_kick(void); 
```
And I thought this was supposed to a *friendly* operating system! :laughing:

keestux commented:
@benpicco I understand the reason to use GCLK2, but I see that RTC and RTT define GCLK2 with source XOSC32K while I have more need for a watchdog that uses OSCULP32K.
SAMD has plenty of GCLKs, why not allocate a dedicated number for WDT? And also give us the freedom to choose the clock source and optionally standby mode, etc.

dylad commented:
> SAMD has plenty of GCLKs, why not allocate a dedicated number for WDT? And also give us the freedom to choose the clock source and optionally standby mode, etc.

Such a configuration can be achieve using a dedicated struct with periph_conf.h

dylad commented:
@benpicco 
Normal mode seems to work for SAML21 but for SAML10 and SAML11 the first test keep resetting forever.

```
2019-05-14 21:14:48,139 - INFO # main(): This is RIOT! (Version: 2019.07-devel-273-g954d1-sam0_wdt)
2019-05-14 21:14:48,140 - INFO # 
2019-05-14 21:14:48,143 - INFO # RIOT Watchdog low-level driver test
2019-05-14 21:14:48,145 - INFO # Normal Mode, 250 ms
2019-05-14 21:14:48,150 - INFO # feeding Watchdog.
2019-05-14 21:14:48,278 - INFO # feeding Watchdog.
2019-05-14 21:14:48,405 - INFO # feeding Watchdog.
2019-05-14 21:14:48,533 - INFO # feeding Watchdog.
2019-05-14 21:14:48,660 - INFO # feeding Watchdog.
2019-05-14 21:14:48,789 - INFO # not feeding the WTD anymore.
2019-05-14 21:14:48,925 - INFO # main(): This is RIOT! (Version: 2019.07-devel-273-g954d1-sam0_wdt)
2019-05-14 21:14:48,925 - INFO # 
2019-05-14 21:14:48,928 - INFO # RIOT Watchdog low-level driver test
2019-05-14 21:14:48,930 - INFO # Normal Mode, 250 ms
2019-05-14 21:14:48,936 - INFO # feeding Watchdog.
2019-05-14 21:14:49,063 - INFO # feeding Watchdog.
2019-05-14 21:14:49,191 - INFO # feeding Watchdog.
2019-05-14 21:14:49,319 - INFO # feeding Watchdog.
2019-05-14 21:14:49,446 - INFO # feeding Watchdog.
2019-05-14 21:14:49,574 - INFO # not feeding the WTD anymore.
2019-05-14 21:14:49,710 - INFO # main(): This is RIOT! (Version: 2019.07-devel-273-g954d1-sam0_wdt)
2019-05-14 21:14:49,711 - INFO # 
2019-05-14 21:14:49,714 - INFO # RIOT Watchdog low-level driver test
2019-05-14 21:14:49,716 - INFO # Normal Mode, 250 ms
2019-05-14 21:14:49,721 - INFO # feeding Watchdog.
2019-05-14 21:14:49,849 - INFO # feeding Watchdog.
2019-05-14 21:14:49,976 - INFO # feeding Watchdog.
2019-05-14 21:14:50,104 - INFO # feeding Watchdog.
2019-05-14 21:14:50,232 - INFO # feeding Watchdog.
2019-05-14 21:14:50,360 - INFO # not feeding the WTD anymore.
2019-05-14 21:14:50,496 - INFO # main(): This is RIOT! (Version: 2019.07-devel-273-g954d1-sam0_wdt)
2019-05-14 21:14:50,497 - INFO # 
2019-05-14 21:14:50,501 - INFO # RIOT Watchdog low-level driver test
2019-05-14 21:14:50,503 - INFO # Normal Mode, 250 ms
2019-05-14 21:14:50,507 - INFO # feeding Watchdog.
2019-05-14 21:14:50,635 - INFO # feeding Watchdog.
2019-05-14 21:14:50,762 - INFO # feeding Watchdog.
2019-05-14 21:14:50,889 - INFO # feeding Watchdog.
2019-05-14 21:14:51,017 - INFO # feeding Watchdog.
2019-05-14 21:14:51,145 - INFO # not feeding the WTD anymore.
```

keestux commented:
Here is my suggestion for clock setup, using GCLK4. The code and comments are from @GabrielNotman in Arduino library Sodaq_wdt.
```
static void _wdt_clock_setup(void)
{
    /* Setup clock provider WDT_GCLK with a 32 source divider
     * GCLK_GENDIV_ID(X), specifies which GCLK we are configuring
     * GCLK_GENDIV_DIV(Y), specifies the clock prescalar / divider
     * If GENCTRL.DIVSEL is set (see further below) the divider 
     * is 2^(Y+1). If GENCTRL.DIVSEL is 0, the divider is simply Y
     * This register has to be written in a single operation
     */
    GCLK->GENDIV.reg = GCLK_GENDIV_ID(WDT_GCLK) | 
                       GCLK_GENDIV_DIV(4);

    /* Configure the GCLK module
     * GCLK_GENCTRL_GENEN, enable the specific GCLK module
     * GCLK_GENCTRL_SRC_OSCULP32K, set the source to the OSCULP32K
     * GCLK_GENCTRL_ID(X), specifies which GCLK we are configuring
     * GCLK_GENCTRL_DIVSEL, specify which prescalar mode we are using
     * Output from this module is 1khz (32khz / 32)
     * This register has to be written in a single operation.
     */
    GCLK->GENCTRL.reg = GCLK_GENCTRL_GENEN | 
                        GCLK_GENCTRL_SRC_OSCULP32K | 
                        GCLK_GENCTRL_ID(WDT_GCLK) | 
                        GCLK_GENCTRL_DIVSEL;
    while (GCLK->STATUS.bit.SYNCBUSY) {}
    
    /* Configure the WDT clock
     * GCLK_CLKCTRL_ID(GCLK_CLKCTRL_ID_WDT), specify the WDT clock
     * GCLK_CLKCTRL_GEN(WDT_GCLK), specify the source from the WDT_GCLK GCLK
     * This register has to be written in a single operation
     */
    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID(GCLK_CLKCTRL_ID_WDT) |
                        GCLK_CLKCTRL_GEN(WDT_GCLK) | 
                        GCLK_CLKCTRL_CLKEN;
    while (GCLK->STATUS.bit.SYNCBUSY) {}
}
```

keestux commented:
The test program does not behave as intended. I have the noinit patch, but still the `magic` is being reset to zero each time. Is that maybe because I have the Arduino bootloader?

benpicco commented:
Yes the Arduino bootloader will [clear the .bss segment](https://github.com/arduino/ArduinoCore-samd/blob/master/bootloaders/zero/board_startup.c#L101).

I think this can be mitigated by placing the `.noinit` section at the end of the RAM, assuming no boot-loader will ever use all the memory (and a heap).
Or placing it after `.bss` and before the heap. I'll prepare a pull request.

I'll include your clock setup code in the next revision when I adapt the driver to the new Watchdog API.

keestux commented:
With #11591 it cycles through the test case. I've rebased this wdt branch ontop of it. Unfortunately it fails the window mode test

```
...
main(): This is RIOT! (Version: 2019.07-devel-295-g52167-b)

RIOT Watchdog low-level driver test
Early Warning Mode, 1000 ms
feeding watchdog.
feeding watchdog.
feeding watchdog.
feeding watchdog.
feeding watchdog.
starving watchdog.
main(): This is RIOT! (Version: 2019.07-devel-295-g52167-b)

RIOT Watchdog low-level driver test
Window Mode, 100 < x < 2500 ms
feeding Watchdog.
main(): This is RIOT! (Version: 2019.07-devel-295-g52167-b)
ERROR - last test failed.
```

MrKevinWeiss commented:
Soft feature freeze is coming up, do you guys think you can still make it (28.06.2019)?

benpicco commented:
I'm on vacation from tomorrow on, so unless someone else wants to pick this up, this will have to wait for the next release.
#11527 needs to be merged first anyway.

MrKevinWeiss commented:
Enjoy your vacation :beach_umbrella:

benpicco commented:
Rebased to make use of #11527.

keestux commented:
@benpicco Great, the test `periph_wdt` is successful on my Sodaq-ONE board now.

Even with my suggested patch to use GCLK4.

benpicco commented:
@dylad do you still see the issue on SAML11?

dylad commented:
Output test result from several boards

SAML11:
```
2019-08-02 21:50:52,076 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:50:52,078 - INFO # Normal Mode, 250 ms
2019-08-02 21:50:52,083 - INFO # kicking Watchdog.
2019-08-02 21:50:52,210 - INFO # kicking Watchdog.
2019-08-02 21:50:52,338 - INFO # kicking Watchdog.
2019-08-02 21:50:52,465 - INFO # kicking Watchdog.
2019-08-02 21:50:52,593 - INFO # kicking Watchdog.
2019-08-02 21:50:52,721 - INFO # not kicking the WTD anymore.
2019-08-02 21:50:52,855 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:50:52,856 - INFO # 
2019-08-02 21:50:52,859 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:50:52,861 - INFO # Normal Mode, 2500 ms
2019-08-02 21:50:52,865 - INFO # kicking Watchdog.
2019-08-02 21:50:54,124 - INFO # kicking Watchdog.
2019-08-02 21:50:55,383 - INFO # kicking Watchdog.
2019-08-02 21:50:56,641 - INFO # kicking Watchdog.
2019-08-02 21:50:57,900 - INFO # kicking Watchdog.
2019-08-02 21:50:59,160 - INFO # not kicking the WTD anymore.
2019-08-02 21:51:00,419 - INFO # board should reset now.
2019-08-02 21:51:00,522 - INFO # board should reset now.
2019-08-02 21:51:00,624 - INFO # board should reset now.
2019-08-02 21:51:00,727 - INFO # board should reset now.
2019-08-02 21:51:00,830 - INFO # board should reset now.
2019-08-02 21:51:00,933 - INFO # board should reset now.
2019-08-02 21:51:01,035 - INFO # board should reset now.
2019-08-02 21:51:01,138 - INFO # board should reset now.
2019-08-02 21:51:01,241 - INFO # board should reset now.
2019-08-02 21:51:01,344 - INFO # board should reset now.
2019-08-02 21:51:01,447 - INFO # board should reset now.
2019-08-02 21:51:01,549 - INFO # board should reset now.
2019-08-02 21:51:01,652 - INFO # board should reset now.
2019-08-02 21:51:01,755 - INFO # board should reset now.
2019-08-02 21:51:01,858 - INFO # board should reset now.
2019-08-02 21:51:01,924 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:51:01,924 - INFO # 
2019-08-02 21:51:01,927 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:51:01,930 - INFO # Window Mode, 100 < x < 2500 ms
2019-08-02 21:51:02,089 - INFO # kicking Watchdog.
2019-08-02 21:51:02,241 - INFO # kicking Watchdog.
2019-08-02 21:51:02,394 - INFO # kicking Watchdog.
2019-08-02 21:51:02,547 - INFO # kicking Watchdog.
2019-08-02 21:51:02,699 - INFO # kicking Watchdog.
2019-08-02 21:51:02,701 - INFO # kicking Watchdog early.
2019-08-02 21:51:02,703 - INFO # You should not see this!
2019-08-02 21:51:06,848 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:51:06,848 - INFO # 
2019-08-02 21:51:06,851 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:51:06,854 - INFO # Early Warning Mode, 1000 ms
2019-08-02 21:51:07,360 - INFO # kicking watchdog.
2019-08-02 21:51:07,865 - INFO # kicking watchdog.
2019-08-02 21:51:08,370 - INFO # kicking watchdog.
2019-08-02 21:51:08,875 - INFO # kicking watchdog.
2019-08-02 21:51:09,381 - INFO # kicking watchdog.
2019-08-02 21:51:09,886 - INFO # starving watchdog.
2019-08-02 21:51:10,396 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:51:10,396 - INFO # 
2019-08-02 21:51:10,399 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:51:10,400 - INFO # Test done.
```

SAML10
```
2019-08-02 21:53:45,517 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:53:45,517 - INFO # 
2019-08-02 21:53:45,521 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:53:45,521 - INFO # Normal Mode, 250 ms
2019-08-02 21:53:45,525 - INFO # kicking Watchdog.
2019-08-02 21:53:45,653 - INFO # kicking Watchdog.
2019-08-02 21:53:45,781 - INFO # kicking Watchdog.
2019-08-02 21:53:45,908 - INFO # kicking Watchdog.
2019-08-02 21:53:46,036 - INFO # kicking Watchdog.
2019-08-02 21:53:46,165 - INFO # not kicking the WTD anymore.
2019-08-02 21:53:46,295 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:53:46,296 - INFO # 
2019-08-02 21:53:46,299 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:53:46,301 - INFO # Normal Mode, 2500 ms
2019-08-02 21:53:46,306 - INFO # kicking Watchdog.
2019-08-02 21:53:47,567 - INFO # kicking Watchdog.
2019-08-02 21:53:48,827 - INFO # kicking Watchdog.
2019-08-02 21:53:50,087 - INFO # kicking Watchdog.
2019-08-02 21:53:51,348 - INFO # kicking Watchdog.
2019-08-02 21:53:52,609 - INFO # not kicking the WTD anymore.
2019-08-02 21:53:53,870 - INFO # board should reset now.
2019-08-02 21:53:53,973 - INFO # board should reset now.
2019-08-02 21:53:54,076 - INFO # board should reset now.
2019-08-02 21:53:54,179 - INFO # board should reset now.
2019-08-02 21:53:54,282 - INFO # board should reset now.
2019-08-02 21:53:54,385 - INFO # board should reset now.
2019-08-02 21:53:54,487 - INFO # board should reset now.
2019-08-02 21:53:54,590 - INFO # board should reset now.
2019-08-02 21:53:54,693 - INFO # board should reset now.
2019-08-02 21:53:54,796 - INFO # board should reset now.
2019-08-02 21:53:54,899 - INFO # board should reset now.
2019-08-02 21:53:55,002 - INFO # board should reset now.
2019-08-02 21:53:55,105 - INFO # board should reset now.
2019-08-02 21:53:55,208 - INFO # board should reset now.
2019-08-02 21:53:55,311 - INFO # board should reset now.
2019-08-02 21:53:55,325 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:53:55,325 - INFO # 
2019-08-02 21:53:55,329 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:53:55,332 - INFO # Window Mode, 100 < x < 2500 ms
2019-08-02 21:53:55,491 - INFO # kicking Watchdog.
2019-08-02 21:53:55,644 - INFO # kicking Watchdog.
2019-08-02 21:53:55,797 - INFO # kicking Watchdog.
2019-08-02 21:53:55,950 - INFO # kicking Watchdog.
2019-08-02 21:53:56,102 - INFO # kicking Watchdog.
2019-08-02 21:53:56,105 - INFO # kicking Watchdog early.
2019-08-02 21:53:56,107 - INFO # You should not see this!
2019-08-02 21:54:00,205 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:54:00,205 - INFO # 
2019-08-02 21:54:00,209 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:54:00,211 - INFO # Early Warning Mode, 1000 ms
2019-08-02 21:54:00,710 - INFO # kicking watchdog.
2019-08-02 21:54:01,210 - INFO # kicking watchdog.
2019-08-02 21:54:01,710 - INFO # kicking watchdog.
2019-08-02 21:54:02,210 - INFO # kicking watchdog.
2019-08-02 21:54:02,709 - INFO # kicking watchdog.
2019-08-02 21:54:03,210 - INFO # starving watchdog.
2019-08-02 21:54:03,713 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:54:03,713 - INFO # 
2019-08-02 21:54:03,716 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:54:03,717 - INFO # Test done.
```

SAML21
```
2019-08-02 21:55:36,945 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:55:36,947 - INFO # Normal Mode, 250 ms
2019-08-02 21:55:36,952 - INFO # kicking Watchdog.
2019-08-02 21:55:37,079 - INFO # kicking Watchdog.
2019-08-02 21:55:37,206 - INFO # kicking Watchdog.
2019-08-02 21:55:37,334 - INFO # kicking Watchdog.
2019-08-02 21:55:37,461 - INFO # kicking Watchdog.
2019-08-02 21:55:37,589 - INFO # not kicking the WTD anymore.
2019-08-02 21:55:37,720 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:55:37,720 - INFO # 
2019-08-02 21:55:37,723 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:55:37,725 - INFO # Normal Mode, 2500 ms
2019-08-02 21:55:37,730 - INFO # kicking Watchdog.
2019-08-02 21:55:38,988 - INFO # kicking Watchdog.
2019-08-02 21:55:40,246 - INFO # kicking Watchdog.
2019-08-02 21:55:41,503 - INFO # kicking Watchdog.
2019-08-02 21:55:42,761 - INFO # kicking Watchdog.
2019-08-02 21:55:44,019 - INFO # not kicking the WTD anymore.
2019-08-02 21:55:45,277 - INFO # board should reset now.
2019-08-02 21:55:45,380 - INFO # board should reset now.
2019-08-02 21:55:45,483 - INFO # board should reset now.
2019-08-02 21:55:45,585 - INFO # board should reset now.
2019-08-02 21:55:45,688 - INFO # board should reset now.
2019-08-02 21:55:45,791 - INFO # board should reset now.
2019-08-02 21:55:45,894 - INFO # board should reset now.
2019-08-02 21:55:45,996 - INFO # board should reset now.
2019-08-02 21:55:46,099 - INFO # board should reset now.
2019-08-02 21:55:46,202 - INFO # board should reset now.
2019-08-02 21:55:46,304 - INFO # board should reset now.
2019-08-02 21:55:46,407 - INFO # board should reset now.
2019-08-02 21:55:46,510 - INFO # board should reset now.
2019-08-02 21:55:46,612 - INFO # board should reset now.
2019-08-02 21:55:46,715 - INFO # board should reset now.
2019-08-02 21:55:46,750 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:55:46,750 - INFO # 
2019-08-02 21:55:46,754 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:55:46,756 - INFO # Window Mode, 100 < x < 2500 ms
2019-08-02 21:55:46,915 - INFO # kicking Watchdog.
2019-08-02 21:55:47,068 - INFO # kicking Watchdog.
2019-08-02 21:55:47,220 - INFO # kicking Watchdog.
2019-08-02 21:55:47,373 - INFO # kicking Watchdog.
2019-08-02 21:55:47,525 - INFO # kicking Watchdog.
2019-08-02 21:55:47,527 - INFO # kicking Watchdog early.
2019-08-02 21:55:47,530 - INFO # You should not see this!
2019-08-02 21:55:51,639 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:55:51,640 - INFO # 
2019-08-02 21:55:51,644 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:55:51,646 - INFO # Early Warning Mode, 1000 ms
2019-08-02 21:55:52,147 - INFO # kicking watchdog.
2019-08-02 21:55:52,648 - INFO # kicking watchdog.
2019-08-02 21:55:53,149 - INFO # kicking watchdog.
2019-08-02 21:55:53,651 - INFO # kicking watchdog.
2019-08-02 21:55:54,152 - INFO # kicking watchdog.
2019-08-02 21:55:54,653 - INFO # starving watchdog.
2019-08-02 21:55:55,157 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:55:55,158 - INFO # 
2019-08-02 21:55:55,160 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:55:55,162 - INFO # Test done.
```

SAME54
```
2019-08-02 21:58:06,313 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:58:06,313 - INFO # 
2019-08-02 21:58:06,316 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:58:06,318 - INFO # Normal Mode, 250 ms
2019-08-02 21:58:06,324 - INFO # kicking Watchdog.
2019-08-02 21:58:06,451 - INFO # kicking Watchdog.
2019-08-02 21:58:06,578 - INFO # kicking Watchdog.
2019-08-02 21:58:06,704 - INFO # kicking Watchdog.
2019-08-02 21:58:06,830 - INFO # kicking Watchdog.
2019-08-02 21:58:06,958 - INFO # not kicking the WTD anymore.
2019-08-02 21:58:07,136 - INFO # board should re�ain(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:58:07,137 - INFO # 
2019-08-02 21:58:07,140 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:58:07,142 - INFO # Normal Mode, 2500 ms
2019-08-02 21:58:07,148 - INFO # kicking Watchdog.
2019-08-02 21:58:08,399 - INFO # kicking Watchdog.
2019-08-02 21:58:09,651 - INFO # kicking Watchdog.
2019-08-02 21:58:10,903 - INFO # kicking Watchdog.
2019-08-02 21:58:12,154 - INFO # kicking Watchdog.
2019-08-02 21:58:13,407 - INFO # not kicking the WTD anymore.
2019-08-02 21:58:14,659 - INFO # board should reset now.
2019-08-02 21:58:14,761 - INFO # board should reset now.
2019-08-02 21:58:14,863 - INFO # board should reset now.
2019-08-02 21:58:14,965 - INFO # board should reset now.
2019-08-02 21:58:15,067 - INFO # board should reset now.
2019-08-02 21:58:15,169 - INFO # board should reset now.
2019-08-02 21:58:15,271 - INFO # board should reset now.
2019-08-02 21:58:15,374 - INFO # board should reset now.
2019-08-02 21:58:15,476 - INFO # board should reset now.
2019-08-02 21:58:15,578 - INFO # board should reset now.
2019-08-02 21:58:15,680 - INFO # board should reset now.
2019-08-02 21:58:15,782 - INFO # board should reset now.
2019-08-02 21:58:15,884 - INFO # board should reset now.
2019-08-02 21:58:15,986 - INFO # board should reset now.
2019-08-02 21:58:16,088 - INFO # board should reset now.
2019-08-02 21:58:16,190 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:58:16,191 - INFO # 
2019-08-02 21:58:16,195 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:58:16,197 - INFO # Window Mode, 100 < x < 2500 ms
2019-08-02 21:58:16,355 - INFO # kicking Watchdog.
2019-08-02 21:58:16,507 - INFO # kicking Watchdog.
2019-08-02 21:58:16,658 - INFO # kicking Watchdog.
2019-08-02 21:58:16,810 - INFO # kicking Watchdog.
2019-08-02 21:58:16,962 - INFO # kicking Watchdog.
2019-08-02 21:58:16,964 - INFO # kicking Watchdog early.
2019-08-02 21:58:16,966 - INFO # You should not see this!
2019-08-02 21:58:21,161 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:58:21,162 - INFO # 
2019-08-02 21:58:21,166 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:58:21,169 - INFO # Early Warning Mode, 1000 ms
2019-08-02 21:58:21,673 - INFO # kicking watchdog.
2019-08-02 21:58:22,179 - INFO # kicking watchdog.
2019-08-02 21:58:22,688 - INFO # kicking watchdog.
2019-08-02 21:58:23,191 - INFO # kicking watchdog.
2019-08-02 21:58:23,697 - INFO # kicking watchdog.
2019-08-02 21:58:24,203 - INFO # starving watchdog.
2019-08-02 21:58:24,734 - INFO # main(): This is RIOT! (Version: 2019.07-devel-1031-g43a86-wdt)
2019-08-02 21:58:24,735 - INFO # 
2019-08-02 21:58:24,738 - INFO # RIOT Watchdog low-level driver test
2019-08-02 21:58:24,738 - INFO # Test done.
```

benpicco commented:
Rebased, now that #11252 is merged.

I dropped my own `tests/periph_wdt` in favor of the one from @fjmolinas 

However, the test will always fail:

```
PASSED target time: 100.0[ms], actual_time: 127.561[ms]
PASSED target time: 500.0[ms], actual_time: 502.736[ms]
PASSED target time: 1000.0[ms], actual_time: 999.896[ms]
FAILED target time: 5000.0[ms], actual_time: 7984.307[ms]
```

The result is the same on samr21-xpro and same54-xpro.
This is because the Watchdog only has 4 bits for window and period respectively.
Only powers of 2 can be configured, so for 5000ms the next power of two (8192/1024Hz = 7938.48 ms) is selected.

fjmolinas commented:
> The result is the same on samr21-xpro and same54-xpro.
> This is because the Watchdog only has 4 bits for window and period respectively.
> Only powers of 2 can be configured, so for 5000ms the next power of two (8192/1024Hz = 7938.48 ms) is selected.

Hmm, this could become an issue for other platforms as well..

We could:
- For now just test for power of two values, or
- Print the theoretically configured time for the setup values and compare against that. This could be a DEBUG option and activate it for the test. e.g: https://github.com/RIOT-OS/RIOT/blob/ab8cdace1206aa271c69b4e1ebcc92c4b6817d89/cpu/stm32_common/periph/wdt.c#L50-L55

What do you think?

fjmolinas commented:
> What do you think?

We can always change the test as well.

benpicco commented:
I think just testing powers of 2 would be the simpler solution.
The watchdog doesn't need *that* much accuracy anyway, so optimizing the test in that direction seems like a wasted effort.

benpicco commented:
With the change to `tests/01-run.py`, the test now passes:

- samr21-xpro
```
PASSED target time: 128[ms], actual_time: 126.273[ms]
PASSED target time: 512[ms], actual_time: 501.609[ms]
PASSED target time: 1024[ms], actual_time: 997.488[ms]
PASSED target time: 8192[ms], actual_time: 7966.684[ms]
TEST PASSED
```

 - same54-xpro
```
PASSED target time: 128[ms], actual_time: 127.315[ms]
PASSED target time: 512[ms], actual_time: 499.86[ms]
PASSED target time: 1024[ms], actual_time: 997.852[ms]
PASSED target time: 8192[ms], actual_time: 7941.416[ms]
TEST PASSED
```

aabadie commented on cpu/sam0_common/periph/wdt.c:
```diff
@@ -0,0 +1,244 @@
+static void _wdt_clock_setup(void)
+{
+/* RTC / RTT will alredy set up GCLK2 as needed */
+#if !defined(USEMODULE_PERIPH_RTC) && !defined(USEMODULE_PERIPH_RTT)
```
```suggestion
#if !defined(MODULE_PERIPH_RTC) && !defined(MODULE_PERIPH_RTT)
```

benpicco replied:
Thank you, good catch!

aabadie commented:
Ran `tests/peripj_wdt` on arduino-zero (samd21), samr30-xpro (saml21) and samr21-xpro: all are working.
Once my comment about the invalid `USEMODULE_PERIPH_WDT` is addressed (you can amend directly), this will be good to go.

benpicco commented:
Had to fix some format string warnings, but I'll still squash.

benpicco commented:
Thank you for testing!

keestux commented:
A pity you didn't consider to use GCLK4. Or shall we do that in a new PR?

benpicco commented:
@keestux the GCLKs need an overhaul anyway, see  #12037
